oauth2_client/client.py

```diff
@@ -29,9 +38,58 @@ class OAuth2Client(OAuth2Session):
         > r = client.get('/api/license/1/detail/')
     """
 
-    def __init__(self, app, token):
-        self.service_host = app.service_host  # used to transform relative URLs to absolute
-        super(OAuth2Client, self).__init__(client_id=app.client_id, token=token)
+    """
+    If token re-fetch happens more than once in this period, raise an exception. Something is wrong,
+    don't spam the provider.
+    """
+    REAUTH_LIMIT_SECONDS = 10
+
+    def __init__(self, token, expired_callback=None):
+        """
+        Create OAuth2Client
+
+        :param token: oauth2_client.model.AccessToken object
+        :param expired_callback: (optional) function accepting one parameter, `Application`. It is
+            called when TokenExpiredError is detected or 403 Forbidden status code received.
+            It should return a valid (not expired) oauth2_client.model.AccessToken object for
+            communication with Resource Owner specified by `Application`.
+            Explanation: The parent class, OAuth2Session already implements automated token refresh,
+            but their solution works only for flows that issue a refresh_token. Both currently
+            supported flows, JWT and Client Credentials don't provide the refresh_token, so we have
+            to handle the refresh ourselves. (Client Credentials MAY issue refresh_token but it
+            doesn't in case of the `oauth2_provider` we use). If the token expiry is detected more
+            than once within `REAUTH_LIMIT_SECONDS` period, `oauthlib.oauth2.TokenExpiredError`
```

Comment:
Is token expiry the only possible cause of the stated condition? I ask because it's important not to mislead future readers with poor naming. Also, have we correctly accounted for timeouts in the server connection? This is another condition where a circuit breaker can stop clients repeatedly beating on the server by returning immediate failure as long as the breaker remains tripped.

Reply:
outdated